src/Compilers/Core/Portable/MetadataReader/MetadataTypeName.cs

```diff
@@ -161,37 +194,58 @@ public string FullName
             }
         }
 
+        /// <inheritdoc cref="NamespaceName"/>
+        public ReadOnlyMemory<char> NamespaceNameMemory
+        {
+            get
+            {
+                if (_namespaceNameMemory.Equals(default(ReadOnlyMemory<char>)))
+                {
+                    Debug.Assert(_fullName != null);
+                    _typeNameMemory = MetadataHelpers.SplitQualifiedName(_fullName, out _namespaceNameMemory);
+                }
+
+                return _namespaceNameMemory;
+            }
+        }
+
         /// <summary>
         /// Namespace name for top level types, empty string for nested types.
         /// </summary>
-        public string NamespaceName
+        public string NamespaceName => _namespaceName ??= NamespaceNameMemory.ToString();
+
+        /// <inheritdoc cref="TypeName"/>
+        public ReadOnlyMemory<char> TypeNameMemory
         {
             get
             {
-                if (_namespaceName == null)
+                if (_typeNameMemory.Equals(default(ReadOnlyMemory<char>)))
                 {
                     Debug.Assert(_fullName != null);
-                    _typeName = MetadataHelpers.SplitQualifiedName(_fullName, out _namespaceName);
+                    _typeNameMemory = MetadataHelpers.SplitQualifiedName(_fullName, out _namespaceNameMemory);
                 }
 
-                return _namespaceName;
+                return _typeNameMemory;
             }
         }
 
         /// <summary>
         /// Name of the type without namespace prefix, but possibly with generic arity mangling present.
         /// </summary>
-        public string TypeName
+        public string TypeName => _typeName ??= TypeNameMemory.ToString();
+
+        /// <inheritdoc cref="UnmangledTypeName"/>
+        public ReadOnlyMemory<char> UnmangledTypeNameMemory
         {
             get
             {
-                if (_typeName == null)
+                if (_unmangledTypeNameMemory.Equals(default(ReadOnlyMemory<char>)))
                 {
-                    Debug.Assert(_fullName != null);
-                    _typeName = MetadataHelpers.SplitQualifiedName(_fullName, out _namespaceName);
+                    // Can't assert anything about arity.  It may have been computed in UnmangledTypeName
```

Comment:
>// Can't assert anything about arity.  It may have been computed in UnmangledTypeName [](http://example.com/codeflow?start=20&length=85)

It looks like the comment is obsolete and we can put the assert back.